File: scripts/autonomy-report.mjs
/** Produce reproducible SIMULATED workflow evidence; never calls a provider or publishes. */
import { writeFile } from "node:fs/promises";
import { runDemo } from "../dist/examples/autonomous/demo.js";
function summarize(comparison) {
  if (!comparison) return null;
  const n = comparison.cases.length;
  return {
    sampleCount: n,
    baselineMean: n
      ? comparison.cases.reduce((s, r) => s + r.baseline, 0) / n
      : null,
    candidateMean: n
      ? comparison.cases.reduce((s, r) => s + r.candidate, 0) / n
      : null,
    metrics: comparison.metrics,
    estimatedServingCost: comparison.estimatedServingCost,
  };
}
const workflows = [];
for (const kind of ["prompt", "models", "decision"])
  for (const path of [
    "accepted",
    "rejected",
    "insufficient",
    "interrupted",
    "regression",
  ]) {
    const report = await runDemo(kind, path);
    workflows.push({
      kind,
      path,
      transitions: report.transitions,
      result: report.result,
      baseline: report.baseline,
      activeVersion: report.activeVersion,
      validation: Object.fromEntries(
        Object.entries(report.validation ?? {}).map(([id, comparison]) => [
          id,
          summarize(comparison),
        ]),
      ),
      audit: summarize(report.audit),
      observations: report.observations,
    });
  }
const evidence = {
  generatedAt: new Date().toISOString(),
  release: "0.3.0-alpha.1",
  runtime: process.version,
  evidence:
    "SIMULATED: deterministic predictions, synthetic labeled examples, in-memory serving registry and simulated observation clock",
  injectedFailures: [
    "Lost apply receipt in interrupted path",
    "Wrong subsequent predictions in regression path",
  ],
  liveModelCalls: 0,
  provesCausalProductionImprovement: false,
  workflows,
};
const output = JSON.stringify(evidence, null, 2) + "\n";
if (process.argv[2]) {
  await writeFile(process.argv[2], output, { flag: "wx" });
  console.log(
    `Wrote ${workflows.length} simulated workflow reports to ${process.argv[2]}`,
  );
} else console.log(output);

File: scripts/autonomy-report.py
"""Retain measured SIMULATED Python workflow reports. No provider calls or publishing."""

import argparse
import asyncio
import importlib.metadata
import json
import platform
import sys
from datetime import UTC, datetime
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1] / "python/examples"))
from autonomous_demo import run_demo


def comparison(value):
    if not value:
        return None
    cases = value["cases"]
    return {
        "sample_count": len(cases),
        "baseline_mean": sum(row["baseline"] for row in cases) / len(cases),
        "candidate_mean": sum(row["candidate"] for row in cases) / len(cases),
        "metrics": value["metrics"],
        "estimated_serving_cost": value["estimated_serving_cost"],
    }


def summarized(report):
    return {
        **report,
        "validation": {
            cid: comparison(value)
            for cid, value in (report["validation"] or {}).items()
        },
        "audit": comparison(report["audit"]),
    }


async def main(args):
    paths = ("accepted", "rejected", "insufficient", "interrupted", "regression")
    reports = []
    for kind in ("prompt", "models", "decision"):
        for path in paths:
            reports.append(summarized(await run_demo(kind, path)))
    optional = []
    if args.sklearn:
        from sklearn_decision_demo import run_sklearn_demo

        for path in paths:
            optional.append(summarized(await run_sklearn_demo(path)))
    evidence = {
        "generated_at": datetime.now(UTC).isoformat(),
        "release": "0.3.0a1",
        "runtime": platform.python_version(),
        "evidence": "SIMULATED: synthetic labels, in-memory registry, simulated observation clock",
        "injected_failures": [
            "Lost apply receipt",
            "Wrong post-deployment predictions",
        ],
        "live_model_calls": 0,
        "proves_causal_production_improvement": False,
        "workflows": reports,
        "optional_sklearn_version": importlib.metadata.version("scikit-learn")
        if args.sklearn
        else None,
        "optional_sklearn": optional,
    }
    with args.output.open("x") as output:
        json.dump(evidence, output, indent=2)
        output.write("\n")
    print(f"Wrote {len(reports) + len(optional)} simulated reports to {args.output}")


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "output", type=Path, help="A new output path; existing reports are preserved"
    )
    parser.add_argument(
        "--sklearn", action="store_true", help="Also exercise the optional ML bridge"
    )
    asyncio.run(main(parser.parse_args()))
